presto-native-execution/presto_cpp/main/runtime-metrics/prometheus-reporter/PrometheusReporter.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "presto_cpp/main/common/Configs.h"
+#include "velox/common/base/Exceptions.h"
+#include "velox/velox/common/base/GTestMacros.h"
+#include "velox/velox/common/base/StatsReporter.h"
+
+namespace facebook::presto::prometheus {
+
+struct StatsInfo {
+  velox::StatType statType;
+  void* metricPtr;
+};
+
+/// Prometheus CPP library exposes following classes:
+/// 1. Registry.
+/// 2. Family.
+/// 3. Specific metric type classes like Counter, Gauge, Histogram etc.
+///
+/// A family of metrics will have the same metric name. For example
+/// http_latency_ms. Different members of a family will have unique labels.
+/// For metric http_latency_ms labels could be {method="GET"}, {method="PUT"},
+/// {method="POST"} etc. Prometheus treats {<metric_name>, [labels]} as unique
+/// metric object.
+class PrometheusReporter : public facebook::velox::BaseStatsReporter {
+  class Impl;
+
+ public:
+  explicit PrometheusReporter(const std::map<std::string, std::string>& labels);
+
+  ~PrometheusReporter() {
+    VLOG(1) << "Deleting prometheus reporter";
+  }
+
+  void registerMetricExportType(const char* key, velox::StatType)
+      const override;
+
+  void registerMetricExportType(folly::StringPiece key, velox::StatType)
+      const override;
+
+  void registerHistogramMetricExportType(
+      const char* key,
+      int64_t bucketWidth,
+      int64_t min,
+      int64_t max,
+      const std::vector<int32_t>& pcts) const override;
+
+  void registerHistogramMetricExportType(
+      folly::StringPiece key,
+      int64_t bucketWidth,
+      int64_t min,
+      int64_t max,
+      const std::vector<int32_t>& pcts) const override;
+
+  void addMetricValue(const std::string& key, size_t value = 1) const override;
+
+  void addMetricValue(const char* key, size_t value = 1) const override;
+
+  void addMetricValue(folly::StringPiece key, size_t value = 1) const override;
+
+  void addHistogramMetricValue(const std::string& key, size_t value)
+      const override;
+
+  void addHistogramMetricValue(const char* key, size_t value) const override;
+
+  void addHistogramMetricValue(folly::StringPiece key, size_t value)
+      const override;
+
+  std::string getSerializedMetrics() const;
+
+ private:
+  // A map of labels assigned to each metric which helps in filtering at client
+  // end.
+  std::shared_ptr<Impl> impl_;
+  // @TODO: may be add StatType::HISTOGRAM to avoid this member.
```

Comment:
nit: Can you clarify this TODO?

Reply:
There is not Statype enum defined for HISTOGRAM and I am using COUNTER as a placeholder here:
https://github.com/prestodb/presto/pull/22360/files#diff-6faf5147ebdc99cf6bb21d2f6bde517e0ee03bf346f59350286a68bf8e912f77R126

I moved this TODO where histograms are registered.

Reply:
I added one on the Velox side https://github.com/facebookincubator/velox/pull/10099